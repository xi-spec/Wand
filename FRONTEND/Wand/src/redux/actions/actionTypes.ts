export default {
  LOAD_PRODUCT_TYPE: 'LOAD_PRODUCT_TYPE',
  LOAD_ALL_PRODUCTS: 'LOAD_ALL_PRODUCTS',
  LOAD_NEW_PRODUCTS: 'LOAD_NEW_PRODUCTS',
  LOAD_PRODUCTS_LIST: 'LOAD_PRODUCTS_LIST',
  LOAD_PRODUCT_DETAIL: 'LOAD_PRODUCT_DETAIL',
  ADD_PRODUCT_TO_CART: ' ADD_PRODUCT_TO_CART',
  INCREASE_QUANTITY_CART: 'INCREASE_QUANTITY_CART',
  DECREASE_QUANTITY_CART: 'DECREASE_QUANTITY_CART',
  DELETE_PRODUCT_CART: 'DELETE_PRODUCT_CART',
  DELETE_ALL_CART: 'DELETE_ALL_CART',
  UPDATE_STOCK_API: 'UPDATE_STOCK_API',
  SEARCH_PRODUCT: 'SEARCH_PRODUCT',
  ADD_PRODUCT_TO_FAVOURITE: 'ADD_PRODUCT_TO_FAVOURITE',
  REGISTER: 'REGISTER',
  REGISTER_ERROR: 'REGISTER_ERROR',
  LOGIN: 'LOGIN',
  LOGIN_ERROR: 'LOGIN_ERROR',
  CLEAN_ISEXIST_STATUS: 'CLEAN_STATUS',
  CLEAN_LOGIN_ERROR: 'CLEAN_LOGIN_ERROR',
  CLEAN_REGISTED: 'CLEAN_REGISTED',
  REMOVE_FAVOURITE: 'REMOVE_FAVOURITE',
  LOGOUT: 'LOGOUT',
  UPDATE_USER: 'UPDATE_USER',
  PAYMENT_SUCCSSFULY: 'PAYMENT_SUCCESSFULY',
  PDATE_STOCK_API: 'PDATE_STOCK_API'
}
;
