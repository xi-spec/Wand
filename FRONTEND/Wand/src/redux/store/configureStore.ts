import reduxImmutableStatInvariant from 'redux-immutable-state-invariant';
import thunk from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import rootReducer from '../reducers';

function configureStore () {
  const composeEnhancers = compose;

  return createStore(
    rootReducer,
    composeEnhancers(applyMiddleware(reduxImmutableStatInvariant(), thunk))
  );
}

export default configureStore();
