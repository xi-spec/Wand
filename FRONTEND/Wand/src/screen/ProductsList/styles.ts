
import { StyleSheet } from 'react-native';
import { SIZE, FONTS, COLORS } from '../../constants/theme';

const styles = StyleSheet.create({
  container: {
    margin: 10,
    paddingTop: 10
  },
  detail: {
    fontSize: SIZE.second,
    fontFamily: FONTS.light,
    marginBottom: 5
  },
  imageContainer: {
    margin: 5,
    paddingBottom: 15
  },
  image: {
    justifyContent: 'center',
    width: 190,
    height: 280,
    borderRadius: 5,
    marginBottom: 10

  },
  cost: {
    fontFamily: FONTS.primary,
    fontSize: SIZE.second,
    color: COLORS.second
  }
});

export default styles
;
