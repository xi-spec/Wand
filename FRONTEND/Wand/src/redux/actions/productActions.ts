import actionTypes from './actionTypes';
import axios from 'axios';
import SERVER from '../../constants/server';
import { Dispatch } from 'redux';
import { Product, ShoppingCart } from '../../models/initiaState';

export function loadProductsType () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_TYPE);
    dispatch({
      type: actionTypes.LOAD_PRODUCT_TYPE,
      data
    });
  };
}

export function loadAllProducts () {
  return async (dispatch : Dispatch) => {
    const { data } = await axios.get(SERVER.URL_PRODUCT);
    dispatch({
      type: actionTypes.LOAD_ALL_PRODUCTS,
      data
    });
  };
}

export function loadProductsList (typeName:string) {
  return {
    type: actionTypes.LOAD_PRODUCTS_LIST,
    typeName
  };
}

export function loadProductDetail (productId:string) {
  return {
    type: actionTypes.LOAD_PRODUCT_DETAIL,
    productId
  };
}

export function addProductToCart (product:Product) {
  return {
    type: actionTypes.ADD_PRODUCT_TO_CART,
    product
  };
}

export function increaseQuantityCart (productId:string) {
  return {
    type: actionTypes.INCREASE_QUANTITY_CART,
    productId
  };
}

export function decreaseQuantityCart (productId:string) {
  return {
    type: actionTypes.DECREASE_QUANTITY_CART,
    productId
  };
}

export function deleteProductCart (productId:string, quantity:number) {
  return {
    type: actionTypes.DELETE_PRODUCT_CART,
    productId,
    quantity
  };
}

export function searchProducts (name:string) {
  return {
    type: actionTypes.SEARCH_PRODUCT,
    name
  };
}

export function updateStockApi (cart:ShoppingCart) {
  return (dispatch : Dispatch) => {
    cart.cartList.forEach(async (product) => {
      const { data } = await axios.put(`${SERVER.URL_PRODUCT}/${product._id}`, { stock: product.stock });
      dispatch({
        type: actionTypes.UPDATE_STOCK_API,
        data
      });
    });
  };
}
