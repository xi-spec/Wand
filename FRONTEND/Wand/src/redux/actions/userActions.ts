import axios from 'axios';
import actionTypes from './actionTypes';
import SERVER from '../../constants/server';
import { User, Product, ShoppingCart } from '../../models/initiaState';
import { Dispatch } from 'redux';

export function addProductToFavourite (product:Product, user:User, favouriteState:boolean) {
  let newUser;
  favouriteState
    ? newUser = { ...user, favourite: [...user.favourite, product] }
    : newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.ADD_PRODUCT_TO_FAVOURITE,
      data
    });
  };
}

export function register (email:string, name:string, password:string) {
  const query = { email, name, password };
  let actionObj;
  return async (dispatch:Dispatch) => {
    try {
      const { data, status } = await axios.post(SERVER.URL_REGISTER, query);
      actionObj = {
        type: actionTypes.REGISTER,
        data,
        status
      };
    } catch (error) {
      actionObj = {
        type: actionTypes.REGISTER_ERROR
      };
    }
    dispatch(
      actionObj
    );
  };
}

export function login (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    let actionObj;
    try {
      const { data } = await axios.post(SERVER.URL_LOGIN, query);
      actionObj = {
        type: actionTypes.LOGIN,
        data

      };
    } catch (error) {
      actionObj = {
        type: actionTypes.LOGIN_ERROR
      };
    }

    dispatch(actionObj);
  };
}

export function logout (email:string, password:string) {
  const query = { email, password };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.post(SERVER.URL_LOGOUT, query);

    dispatch({
      type: actionTypes.LOGOUT,
      data
    });
  };
}

export function cleanIsExistStatus () {
  return {
    type: actionTypes.CLEAN_ISEXIST_STATUS
  };
}

export function cleanLoginError () {
  return {
    type: actionTypes.CLEAN_LOGIN_ERROR
  };
}

export function cleanRegisted () {
  return {
    type: actionTypes.CLEAN_REGISTED
  };
}

export function removeFavourite (product:Product, user:User) {
  const newUser = { ...user, favourite: user.favourite.filter(elem => elem._id !== product._id) };

  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, newUser);
    dispatch({
      type: actionTypes.REMOVE_FAVOURITE,
      data
    });
  };
}

export function updateAddress (street:string,
  city:string, country:string, postcode:string, user:User) {
  const query = { street, city, country, postcode };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { address: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function updateCard (cardNumber:string, securtyCode:string, cardholder:string, expiryDate:string, user:User) {
  const query = { cardNumber, securtyCode, cardholder, expiryDate };
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { card: query });
    dispatch({
      type: actionTypes.UPDATE_USER,
      data
    });
  };
}

export function paymentSuccessfuly (user:User, cart:ShoppingCart) {
  return async (dispatch:Dispatch) => {
    const { data } = await axios.put(`${SERVER.URL_USER}/${user._id}`, { orderHistory: [...user.orderHistory, cart] });
    dispatch({
      type: actionTypes.PAYMENT_SUCCSSFULY,
      data
    });
  };
}
