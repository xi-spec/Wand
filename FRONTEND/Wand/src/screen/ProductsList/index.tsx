import React, { useEffect } from 'react';
import {
  FlatList,
  TouchableOpacity,
  Text,
  Image
} from 'react-native';
import styles from './styles';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { loadProductsList } from '../../redux/actions/productActions';
import { Product } from '../../models/initiaState';
import setCostValue2Decimals from '../../utils/setCostValue2Decimals';

function ProductsList ({ route, actions, productsList, navigation }:
  {route:{params:{typeName:string}},
   actions:any,
   productsList:Product[],
   navigation:{navigate:Function}}) {
  const { typeName } = route?.params;
  useEffect(() => {
    actions.loadProductsList(typeName);
  }, []);

  return (
        <FlatList
        showsVerticalScrollIndicator={false}
        bounces={false}
        contentContainerStyle={styles.container}
        data={productsList}
        numColumns={2}
        keyExtractor={(item) => item._id}
        renderItem={({ item }) =>
        <TouchableOpacity
        style={styles.imageContainer}
        key={item.name}
        testID='productsList'
        onPress={() => navigation.navigate('Detail', { productId: item._id })}>
      <Image source={{ uri: item.images.alls[0] }}
      style={styles.image}>
        </Image>
        <Text style={styles.detail}>{item.name}</Text>
        <Text style={styles.cost}>{`${setCostValue2Decimals(item.cost.value)} ${item.cost.currency}`}</Text>
      </TouchableOpacity>
        }>
      </FlatList>

  );
}

function mapStateToProps ({ products: { productsList } }) {
  return { productsList };
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      loadProductsList
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductsList)
;
