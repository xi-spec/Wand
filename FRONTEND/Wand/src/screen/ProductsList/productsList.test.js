import React from 'react';
import ProductsList from './index';
import { render, fireEvent, act } from '@testing-library/react-native';
import * as actions from '../../redux/actions/productActions';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';

jest.mock('../../redux/actions/productActions');

describe('Given a ProductsList component', () => {
  const mockStore = configureStore([thunk]);
  const store = mockStore({
    products: {
      productsList: [{
        name: 'be happy 1',
        cost: { value: 2, currency: 'euro' },
        images: { alls: ['url'] }
      }]
    }
  });
  let route;
  beforeEach(() => {
    jest.spyOn(actions, 'loadProductsList').mockReturnValueOnce({ type: '', typeName: 'Rings' });
    route = {
      params: { typeName: 'Rings' }
    };
  });
  describe('When is invoked', () => {
    test('Then it is renders correctly', () => {
      const tree = render(<Provider store={store}><ProductsList route ={route} /></Provider>);
      expect(tree).toMatchSnapshot();
    });
    test('Then actions.loadProductsList is invoked', () => {
      render(<Provider store={store}><ProductsList route ={route} /></Provider>);
      expect(actions.loadProductsList).toHaveBeenCalled();
    });
  });
  describe('When is press in testId productsList', () => {
    test('Then navigation.navigate is invoked', () => {
      const navigation = {
        navigate: jest.fn()
      };

      const { getByTestId } = render(<Provider store={store}><ProductsList navigation={navigation} route ={route} /></Provider>);
      const touch = getByTestId('productsList');

      act(() => {
        fireEvent.press(touch);
      });

      expect(navigation.navigate).toHaveBeenCalled();
    });
  });
});
