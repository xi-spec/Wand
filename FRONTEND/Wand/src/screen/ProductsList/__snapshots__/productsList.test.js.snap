// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a ProductsList component When is invoked Then it is renders correctly 1`] = `
<RCTScrollView
  bounces={false}
  contentContainerStyle={
    Object {
      "margin": 10,
      "paddingTop": 10,
    }
  }
  data={
    Array [
      Object {
        "cost": Object {
          "currency": "euro",
          "value": 2,
        },
        "images": Object {
          "alls": Array [
            "url",
          ],
        },
        "name": "be happy 1",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "margin": 5,
              "opacity": 1,
              "paddingBottom": 15,
            }
          }
          testID="productsList"
        >
          <Image
            source={
              Object {
                "uri": "url",
              }
            }
            style={
              Object {
                "borderRadius": 5,
                "height": 280,
                "justifyContent": "center",
                "marginBottom": 10,
                "width": 190,
              }
            }
          />
          <Text
            style={
              Object {
                "fontFamily": "RalewayLight",
                "fontSize": 14,
                "marginBottom": 5,
              }
            }
          >
            be happy 1
          </Text>
          <Text
            style={
              Object {
                "color": "#F3603F",
                "fontFamily": "MontserratRegular",
                "fontSize": 14,
              }
            }
          >
            2.00 euro
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
